### PR TITLE
refactor(protocol): restructure solidity code to match compilation targets

### PR DESCRIPTION
**Pull request by dantaik**
## Summary
Prior to this PR, both L1 and L2 code were compiled using the same Solidity compilation target (`cancun`), which worked for now but is incompatible with future Solidity versions.

This PR reorganizes the Solidity codebase into three subdirectories:
- `shared`: Contains contracts used by both L1 and L2.
- `layer1`: Contains L1-specific contracts.
- `layer2`: Contains L2-specific contracts.

For L1 compilation, the source directory is now `contracts/layer1`, with `contracts/shared` automatically included. The EVM target for L1 is specified as `cancun`. Similarly, L2 uses `contracts/layer2` as the source and `out/layer2` as the destination with its EVM target set to `shanghai`. The build directories for L1 and L2 are also separately defined.

You can use commands to test contracts for layer1 and layer2 separately now:

- `pnpm compile:l1`, `pnpm compile:l2`, or `pnpm compile` to compile all files.
- `pnpm test:l1`,  `pnpm test:l2`, , or `pnpm test` to run all tests.

Note that:

- Going forward, it's critical to ensure that the contracts in the `shared` directory remain compatible with both the `layer1` and `layer2` codebases.
- Minimal code changes in `EssentialContract.sol` and `Bridge.sol`. **Please focus on reviewing these specific changes**.

This is a large PR, but the majority of the changes involve the directory restructuring rather than the code itself.

### CODE REVIEW

File: .github/workflows/protocol.yml [hunk 1]
```diff
@@ -34,30 +34,22 @@ jobs:
...
 
       - name: Commit contract layout table
         uses: stefanzweifel/git-auto-commit-action@v5
         with:
-          commit_message: "forge fmt & update contract layout table"
+          commit_message: "forge fmt & update contract layout tables"
 
       - name: Generate genesis
         working-directory: ./packages/protocol
         run: pnpm test:genesis
```
Comment (YoGhurt111) [inline .github/workflows/protocol.yml:46]: Should it be `pnpm test:genesis:l2`? Always use this command to generate genesis on devnet

Comment (dantaik) [reply .github/workflows/protocol.yml:46]: It's OK to call it test:genesis

**Comment by dantaik** [comment]
I'll only merge this if client/node has no issues for integration @davidtaikocha @YoGhurt111

**Review by davidtaikocha — DISMISSED** [review]

**Review by davidtaikocha — APPROVED** [review]

File: .gitignore [hunk 2]
```diff
@@ -116,7 +116,6 @@ __pycache__/
 
 # VSCode
 .vscode/launch.json
```
Comment (davidtaikocha) [inline .gitignore:118]:
```suggestion
.vscode/launch.json

```

File: .gitignore [hunk 3]
```diff
@@ -116,7 +116,7 @@ __pycache__/
...
-packages/protocol/config.js
+
```
Comment (davidtaikocha) [inline .gitignore:119]:
```suggestion
```

**Review by YoGhurt111 — DISMISSED** [review]

**Comment by dantaik** [comment]
All checks pass. I will enable auto merging, please approve only you think it's ok to merge.

**Review by davidtaikocha — APPROVED** [review]

**Review by davidtaikocha — DISMISSED** [review]

**Review by YoGhurt111 — APPROVED** [review]

**Review by davidtaikocha — APPROVED** [review]

File: packages/taiko-client/integration_test/deploy_l1_contract.sh [hunk 4]
```diff
@@ -6,7 +6,7 @@ source scripts/common.sh
 source integration_test/l1_env.sh
 
 cd ../protocol &&
-  forge script script/DeployOnL1.s.sol:DeployOnL1 \
+  forge script script/layer1/DeployProtocolOnL1.s.sol:DeployProtocolOnL1 \
```
Comment (dantaik) [inline packages/taiko-client/integration_test/deploy_l1_contract.sh:9]:
```suggestion
  forge script script/DeployOnL1.s.sol:DeployOnL1 \
```

**Review by davidtaikocha — APPROVED** [review]

**Pull request merged**